client/rest/src/main/java/org/elasticsearch/client/RestClientActions.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.client;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.nio.protocol.HttpAsyncResponseConsumer;
+
+/**
+ * Actions that can be taken on a {@link RestClient} or the stateless views
+ * returned by {@link RestClientActions#withNodeSelector withNodeSelector}.
+ */
+public interface RestClientActions {
```

Comment:
we talked about this new way of sending requests with @nik9000 and we said that we will look into changing the existing performRequest methods to accept a `Request` object, so that we can trim down the different performRequest methods that we have to two: sync and async. I feel like this is going to be more familiar to most of our users and something that we should have done from day one.